include/hipSYCL/glue/generic/hiplike/hiplike_kernel_launcher.hpp

```diff
@@ -420,6 +420,11 @@ determine_reduction_stages(sycl::range<Dimensions> global_size,
   sycl::range<Dimensions> current_num_groups = num_groups;
   sycl::range<Dimensions> current_num_work_items = global_size;
 
+  for (int i=0; i < Dimensions; ++i)
+    // Check if local_size == 1, this would cause an infinite loop
+    if (local_size[i] == 1)
+      local_size[i] = 128;
```

Comment:
So, if Dimensions is greater than 1, we end up with (128, 128, ...) as work group size. This probably will break on CUDA and HIP, since those devices typically support at most 1024 elements in a work group.

Wouldn't it be sufficient to execute the reduction kernel always as a one-dimensional thing?

Reply:
Yes that makes sense, I didn't really think about what happens if `Dimensions > 1`. But I don't quite understand what you mean by executing the kernel as a 1D thing

Reply:
I meant just executing the kernel with a 1 dimensional parallel iteration space, i.e. using as a group size something like `(size, 1, 1)` if `Dimensions == 3`.
(Please check whether it needs to be `(size, 1, 1)` or `(1, 1, size)` - SYCL works in terms of `(1, 1, size)`, but CUDA/HIP in terms of (`size, 1, 1)`, so we flip the indices somewhere in the kernel launcher).

Reply:
I'm not sure if I understand. 

Say we have a reduction over a `nd_range<2>({N, M}, {1, 16})`.  Then in the first dimension we still have the problem that computing the reduction stage sizes ends in an infinite loop (because the number of groups in the first step will be computed as `num_groups[0] = N / local_size[0]`, then this will again be divided by `local_size[0]` etc. until `num_groups[0] == 1` but `local_size[0] == 1` so always `num_groups[0] == N`).

Reply:
That's true, but I think this problem only affects when we launch dedicated reduction kernels after the user-provided kernels here:
https://github.com/illuhad/hipSYCL/blob/56df01807eeb81cde76c4e0d6266d78e365ccf24/include/hipSYCL/glue/generic/hiplike/hiplike_kernel_launcher.hpp#L852

I think these could be run as 1D, since all they have to do is collect values from previous kernels and reduce them further.
In fact, looking at the code, they seem to be invoked as 1D already - so can we maybe just somehow remove the `Dimensions > 1` code paths in `determine_reduction_stages()`?.
I think all that code actually needs to care about is the total group size (`x * y * z`) so that it knows how many items will be reduced to one.

Reply:
Okay, I see, makes sense :)

I've changed it so that ranges of the form `(N,M,L)` are transformed to (N*M*L,1,1)` (or `(N,M)` to `(N*M,1)` in case of 2D) before computing the reduction stages.